### PR TITLE
[InstCombine]: Handle known negative variables with KnownBits

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -1491,12 +1491,18 @@ Instruction *InstCombinerImpl::visitUDiv(BinaryOperator &I) {
   }
 
   // Op0 / C where C is large (negative) --> zext (Op0 >= C)
-  // TODO: Could use isKnownNegative() to handle non-constant values.
   Type *Ty = I.getType();
   if (match(Op1, m_Negative())) {
     Value *Cmp = Builder.CreateICmpUGE(Op0, Op1);
     return CastInst::CreateZExtOrBitCast(Cmp, Ty);
   }
+
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:1499]: The `m_Negative` code is now unnecessary.

File: llvm/test/Transforms/InstCombine/udiv-simplify.ll [hunk 2]
```diff
@@ -98,6 +98,18 @@ define i8 @udiv_demanded_low_bits_set(i8 %a) {
...
 }
 
+define i8 @udiv_large_value(i8 %a) {
+; CHECK-LABEL: @udiv_large_value(
+; CHECK-NEXT:    [[TMP1:%.*]] = and i8 [[A:%.*]], -76
+; CHECK-NEXT:    [[TMP2:%.*]] = icmp eq i8 [[TMP1]], -76
+; CHECK-NEXT:    [[U:%.*]] = zext i1 [[TMP2]] to i8
+; CHECK-NEXT:    ret i8 [[U]]
+;
+  %o = or i8 %a, 180
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/udiv-simplify.ll:108]: imo make the constant `128` so the verification of this transform is the most general case.

**Comment by nikic** [comment]
@dtcxzyw Does this have any impact? If not I'm going to drop the TODO instead.

**Comment by goldsteinn** [comment]
> @dtcxzyw Does this have any impact? If not I'm going to drop the TODO instead.

+1

**Pull request closed without merging**